# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [2]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [76]:
pd.set_option('display.max_column',None)
df = pd.read_csv('step3_output.csv', index_col=0)
print(df.shape)
df.head()

(171, 26)


Name       state  summit_elev  vertical_drop  trams  \
2         Hilltop Ski Area      Alaska       2090.0          294.0    0.0   
4      Sunrise Park Resort     Arizona      11100.0         1800.0    0.0   
8   Boreal Mountain Resort  California       7700.0          500.0    0.0   
9              Dodge Ridge  California       8200.0         1600.0    0.0   
10        Donner Ski Ranch  California       8012.0          750.0    0.0   

    fastEight  fastSixes  fastQuads  quad  triple  double  surface  \
2         0.0        0.0        0.0   0.0     1.0     0.0      2.0   
4         0.0        0.0        1.0   2.0     3.0     1.0      0.0   
8         0.0        0.0        1.0   1.0     3.0     1.0      2.0   
9         0.0        0.0        0.0   1.0     2.0     5.0      4.0   
10        0.0        0.0        0.0   0.0     1.0     5.0      2.0   

    total_chairs  Runs  TerrainParks  LongestRun_mi  SkiableTerrain_ac  \
2            3.0  13.0           1.0            1.0               30.0   
4            7.0  65.0           2.0            1.2              800.0   
8            8.0  33.0           6.0            1.0              380.0   
9           12.0  67.0           5.0            2.0              862.0   
10           8.0  52.0           2.0            1.5              505.0   

    Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  \
2        30.000000        150.000000       36.0             69.0   
4        80.000000        115.000000       49.0            250.0   
8       200.000000        150.000000       54.0            400.0   
9       469.052632        151.578947       69.0            350.0   
10       60.000000        163.000000       82.0            400.0   

    AdultWeekday  AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
2           30.0      34.00000              152.0            30.0         0  
4           74.0      78.00000              104.0            80.0         1  
8           49.0      64.16681              150.0           200.0         1  
9           78.0      78.00000              140.0             0.0         1  
10          75.0      75.00000              170.0             0.0         1

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [100]:
state_dummies = pd.get_dummies(df['state'], drop_first=True)
df_states = pd.concat([df, state_dummies], axis=1)
df_states.drop('state', axis=1, inplace=True)

In [101]:
print(df_states.shape)
df_states.head()

(171, 58)


Name  summit_elev  vertical_drop  trams  fastEight  \
2         Hilltop Ski Area       2090.0          294.0    0.0        0.0   
4      Sunrise Park Resort      11100.0         1800.0    0.0        0.0   
8   Boreal Mountain Resort       7700.0          500.0    0.0        0.0   
9              Dodge Ridge       8200.0         1600.0    0.0        0.0   
10        Donner Ski Ranch       8012.0          750.0    0.0        0.0   

    fastSixes  fastQuads  quad  triple  double  surface  total_chairs  Runs  \
2         0.0        0.0   0.0     1.0     0.0      2.0           3.0  13.0   
4         0.0        1.0   2.0     3.0     1.0      0.0           7.0  65.0   
8         0.0        1.0   1.0     3.0     1.0      2.0           8.0  33.0   
9         0.0        0.0   1.0     2.0     5.0      4.0          12.0  67.0   
10        0.0        0.0   0.0     1.0     5.0      2.0           8.0  52.0   

    TerrainParks  LongestRun_mi  SkiableTerrain_ac  Snow Making_ac  \
2            1.0            1.0               30.0       30.000000   
4            2.0            1.2              800.0       80.000000   
8            6.0            1.0              380.0      200.000000   
9            5.0            2.0              862.0      469.052632   
10           2.0            1.5              505.0       60.000000   

    daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
2         150.000000       36.0             69.0          30.0      34.00000   
4         115.000000       49.0            250.0          74.0      78.00000   
8         150.000000       54.0            400.0          49.0      64.16681   
9         151.578947       69.0            350.0          78.0      78.00000   
10        163.000000       82.0            400.0          75.0      75.00000   

    projectedDaysOpen  NightSkiing_ac  clusters  Arizona  California  \
2               152.0            30.0         0        0           0   
4               104.0            80.0         1        1           0   
8               150.0           200.0         1        0           1   
9               140.0             0.0         1        0           1   
10              170.0             0.0         1        0           1   

    Colorado  Connecticut  Idaho  Illinois  Indiana  Iowa  Maine  Maryland  \
2          0            0      0         0        0     0      0         0   
4          0            0      0         0        0     0      0         0   
8          0            0      0         0        0     0      0         0   
9          0            0      0         0        0     0      0         0   
10         0            0      0         0        0     0      0         0   

    Massachusetts  Michigan  Minnesota  Missouri  Montana  Nevada  \
2               0         0          0         0        0       0   
4               0         0          0         0        0       0   
8               0         0          0         0        0       0   
9               0         0          0         0        0       0   
10              0         0          0         0        0       0   

    New Hampshire  New Jersey  New Mexico  New York  North Carolina  Ohio  \
2               0           0           0         0               0     0   
4               0           0           0         0               0     0   
8               0           0           0         0               0     0   
9               0           0           0         0               0     0   
10              0           0           0         0               0     0   

    Oregon  Pennsylvania  Rhode Island  South Dakota  Utah  Vermont  Virginia  \
2        0             0             0             0     0        0         0   
4        0             0             0             0     0        0         0   
8        0             0             0             0     0        0         0   
9        0             0             0             0     0        0         0   
10 

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [103]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_states.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df_states.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [109]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [110]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [111]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [112]:
y_test

array([76.        , 70.        , 55.        , 47.        , 45.        ,
       44.        , 49.        , 56.        , 38.        , 57.        ,
       50.        , 75.        , 44.        , 45.        , 52.        ,
       70.        , 68.        , 43.        , 72.        , 45.        ,
       49.        , 59.        , 92.        , 70.        , 79.        ,
       89.        , 65.        , 35.        , 65.        , 48.        ,
       64.16681004, 72.        , 64.16681004, 64.16681004, 35.        ,
       79.        , 49.        , 48.        , 59.        , 49.        ,
       45.        , 64.16681004, 75.        ])

In [113]:
y_pred

array([68.82812243, 63.55232466, 55.75971748, 40.11884034, 48.90740297,
       42.27743489, 61.75840318, 67.67729965, 40.91361368, 59.13407225,
       53.32492189, 69.95198544, 46.58530793, 46.00583081, 54.52072543,
       71.49329355, 53.87206037, 53.77746882, 71.67538874, 45.36124071,
       50.56945834, 57.07957934, 83.50242529, 66.82036787, 70.90743305,
       86.85822018, 48.08094262, 37.56230218, 73.51237922, 51.57662313,
       48.85759196, 83.37712033, 54.42063326, 63.39284581, 54.15848271,
       76.31276245, 54.57169858, 39.71263857, 56.30823917, 47.95750705,
       51.63217502, 54.61754552, 68.8460828 ])

In [114]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [115]:
print(f'Explained variance score:  {explained_variance_score(y_test,y_pred)}')
print(f'Mean absolute error:       {mean_absolute_error(y_test, y_pred)}')

Explained variance score:  0.7108353714508809
Mean absolute error:       5.904553632166868


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [116]:
print(f'The linear regression model intercept is: {lm.intercept_}')

The linear regression model intercept is: 16.839848567852663


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [117]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
West Virginia     16.364975
Rhode Island      14.405107
Massachusetts     12.645095
Maine             10.941464
Virginia          10.434404
Montana            9.975327
Wyoming            9.773201
North Carolina     9.720326
New Hampshire      9.001503
Colorado           8.641765

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [118]:
df.head(2)

Name    state  summit_elev  vertical_drop  trams  fastEight  \
2     Hilltop Ski Area   Alaska       2090.0          294.0    0.0        0.0   
4  Sunrise Park Resort  Arizona      11100.0         1800.0    0.0        0.0   

   fastSixes  fastQuads  quad  triple  double  surface  total_chairs  Runs  \
2        0.0        0.0   0.0     1.0     0.0      2.0           3.0  13.0   
4        0.0        1.0   2.0     3.0     1.0      0.0           7.0  65.0   

   TerrainParks  LongestRun_mi  SkiableTerrain_ac  Snow Making_ac  \
2           1.0            1.0               30.0            30.0   
4           2.0            1.2              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
2             150.0       36.0             69.0          30.0          34.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
2              152.0            30.0         0  
4              104.0            80.0         1

In [119]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

X = df.drop(['Name','state','AdultWeekend'], axis=1)
y = df.AdultWeekend 

scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [120]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [121]:
# Model 2 
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

# create the prediction
y_pred = model.predict(X_test)

In [122]:
# Review the model metrics
print(f'Explained variance score:  {explained_variance_score(y_test,y_pred)}')
print(f'Mean absolute error:       {mean_absolute_error(y_test, y_pred)}')
print(f'The linear regression model intercept is: {lm.intercept_}')

Explained variance score:  0.7342384977752414
Mean absolute error:       5.500970711020954
The linear regression model intercept is: 56.82330114866418


In [123]:
# Look at the most weighing features
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
AdultWeekday         10.202972
summit_elev           3.075028
SkiableTerrain_ac     2.388004
clusters              2.358982
daysOpenLastYear      2.180618
triple                1.874711
vertical_drop         1.775797
Runs                  1.706097
surface               1.393971
projectedDaysOpen     1.355468

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [125]:
# 'base_elev' column was dropped in the previous step3
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

X = df.drop(['Name','state','AdultWeekend','summit_elev'], axis=1)
y = df.AdultWeekend 

scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [126]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [127]:
# Model 2 
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

# create the prediction
y_pred = model.predict(X_test)

In [128]:
# Review the model metrics
print(f'Explained variance score:  {explained_variance_score(y_test,y_pred)}')
print(f'Mean absolute error:       {mean_absolute_error(y_test, y_pred)}')
print(f'The linear regression model intercept is: {lm.intercept_}')

Explained variance score:  0.7338777613923764
Mean absolute error:       5.682970469049308
The linear regression model intercept is: 56.89353094368575


In [129]:
# Look at the most weighing features
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
AdultWeekday         10.199130
SkiableTerrain_ac     2.977775
daysOpenLastYear      1.938667
averageSnowfall       1.836155
Runs                  1.793607
clusters              1.739456
vertical_drop         1.628720
triple                1.543080
surface               1.142156
NightSkiing_ac        1.139455

### Additional model... update after reviewing results from previous model
#### Model 4

In [130]:
# load dataframe that ignored removing outliers (there were a few exceptions explained in step3)
df2 = pd.read_csv('step3_output_dfonly.csv', index_col=0)

In [131]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

X = df2.drop(['Name','state','AdultWeekend','summit_elev'], axis=1)
y = df2.AdultWeekend 

scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [132]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [133]:
# Load all important libraries from sklearn
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [134]:
# Model 2 
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

# create the prediction
y_pred = model.predict(X_test)

In [135]:
# Review the model metrics
print(f'Explained variance score:  {explained_variance_score(y_test,y_pred)}')
print(f'Mean absolute error:       {mean_absolute_error(y_test, y_pred)}')
print(f'The linear regression model intercept is: {lm.intercept_}')

Explained variance score:  0.8196870402850571
Mean absolute error:       6.529629100131694
The linear regression model intercept is: 64.07544480678818


In [136]:
# Look at the most weighing features
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
AdultWeekday         20.534312
daysOpenLastYear      2.377590
fastQuads             1.439772
averageSnowfall       1.405344
quad                  1.121443
SkiableTerrain_ac     1.112249
projectedDaysOpen     1.011300
surface               1.002937
NightSkiing_ac        0.893359
total_chairs          0.864758

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.711 | 5.905 |-|
| Model 2. | 0.734 | 5.501 |'state'|
| Model 3. | 0.734 | 5.683 |'state','summit_elev','base_elev'|
| Model 4. | 0.820 | 6.530 |'state','summit_elev','base_elev' - ignored outlier removal|

**Model Selection:** Model 3 and 4 will be reviewed.